nth(), first() and last()

romainfrancois opened the pull request:
closes #35

This is for now similar to the versions in dplyr except that it does not have `order_by=`. Should it ?

Also wondering: 
 - should this keep supporting negative values (to count from the right) ?
 - out of bounds return the default, should that rather be an error ?

Apart from that, this is mostly a wrapper around `vec_slice()`: 

``` r
library(funs, warn.conflicts = FALSE)

nth(letters, 1)
#> [1] "a"
nth(letters, -1)
#> [1] "z"
nth(letters, 30)
#> [1] NA
nth(character(), 2)
#> [1] NA

df <- data.frame(x = 1:10, y = letters[1:10])
nth(df, 3)
#>   x y
#> 1 3 c
nth(df, -2)
#>   x y
#> 1 9 i
nth(df, 11)
#>    x    y
#> 1 NA <NA>

mat <- matrix(1:20, ncol = 4)
nth(mat, 2)
#>      [,1] [,2] [,3] [,4]
#> [1,]    2    7   12   17
nth(mat, -1)
#>      [,1] [,2] [,3] [,4]
#> [1,]    5   10   15   20

lst <- list(1, letters, TRUE)
nth(lst, 1)
#> [[1]]
#> [1] 1
nth(lst, 4)
#> [[1]]
#> NULL
```

<sup>Created on 2020-09-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

romainfrancois committed:
initial nth(), first() and last()

romainfrancois committed:
initial nth(), first() and last()

romainfrancois committed:
use vec_size() instead of length()

romainfrancois committed:
use vec_size() instead of length()

romainfrancois committed:
get vec_size() earlier

romainfrancois committed:
get vec_size() earlier

hadley commented:
I think we have to keep supporting negative values in order to not break existing dplyr code. And I'm pretty sure we have to return the default value for OOB values, since that's the primary value of these functions over `[[`.

@DavisVaughan do you have any thoughts on whether we should support `order_by` here?

DavisVaughan commented on R/nth.R:
```diff
@@ -0,0 +1,38 @@
+    return(nth_default(x, default))
+  }
+
+  if (!is_integerish(n)) {
```
```suggestion
  if (!is_integerish(n, n = 1L, finite = TRUE)) {
```

I think we need the `n = 1L` check, but also consider `finite = TRUE` to prevent `nth(x, Inf)`?

romainfrancois replied:
yes thanks

DavisVaughan reviewed (commented):
I don't think I care too much about these functions supporting `order_by`. I do think lag/lead should, to be compatible with the ilag/ilead ideas in https://github.com/tidyverse/funs/issues/46

---

Side note - Consequences of using `vec_slice()` here vs the `[[` equivalent of `vec_slice2()`:
- Lists are not unpacked
- `vec_names()` are kept on the result (names on vectors, row names on dfs/arrays)

We should probably document these if we still believe that `vec_slice()` is the right thing to use here as mentioned in
https://github.com/tidyverse/funs/issues/35#issuecomment-564764617, along with the invariant there that this always returns something with size 1.

romainfrancois committed:
Update R/nth.R

Co-authored-by: Davis Vaughan <davis@rstudio.com>

romainfrancois committed:
using testthat 3rd edition (#61)

* declare using testthat 3rd edition

* typo

* using expect_snapshot()

* Add NEWS.md

romainfrancois committed:
update

Merge branch 'nth' of https://github.com/romainfrancois/funs into nth

# Conflicts:
#	R/nth.R

romainfrancois committed:
x_size -> size

DavisVaughan commented on R/nth.R:
```diff
@@ -0,0 +1,38 @@
+nth <- function(x, n, default = NA) {
+  x_size <- vec_size(x)
+
+  if (x_size == 0) {
+    return(nth_default(x, default))
+  }
```
I think this is nicely captured by `n == 0 || n > x_size || n < -x_size` and might not be required

romainfrancois replied:
Thanks

romainfrancois committed:
initial documentation for first(), last(), nth()

romainfrancois committed:
remove unnecessary special case

romainfrancois committed:
tests

romainfrancois committed:
assignment versions of nth(), first(), last()

closes #20

romainfrancois committed:
param value

DavisVaughan commented on R/nth.R:
```diff
@@ -0,0 +1,107 @@
+    abort("`n` must be a single integer.")
+  }
+
+  if (n == 0 || n > size || n < -size) {
+    abort("`n` is out of bounds")
+  }
```
If `nth<-` is supposed to be lax like `nth()`, wouldn't this just return `x` unmodified?

If so, I would also move it after the assertions and casting of `value`, to ensure those checks are run before we early exit

romainfrancois replied:
Maybe with a warning ?

DavisVaughan replied:
That seems reasonable to me

DavisVaughan commented on R/nth.R:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  vec_assert(value, size = 1L)
+  params <- vec_cast_common(x = x, value = value)
```
I would have expected that the result of calling `nth<-` is type stable on `x`.

So `value <- vec_cast(value, x, x_arg = "value", to_arg = "x")`, rather than taking the common type.

romainfrancois replied:
yeah that's probably better

romainfrancois committed:
nth<- type stable on x

romainfrancois committed:
return x unmodified when n is out of bounds, so that nth<-() is lax

romainfrancois committed:
explicitly set to use strings

romainfrancois committed:
warn on nth(x, n=<out of bounds>) <-

romainfrancois commented:
Is this going to be annoying ? 

``` r
library(funs, warn.conflicts = FALSE)

x <- 1:4
nth(x, 6) <- 5
#> Warning: Unchanged `x` as `n` is out of bounds
x
#> [1] 1 2 3 4
```

<sup>Created on 2021-05-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

lionel- commented:
I think we generally prefer errors to warnings? I.e. the behaviour should either be silent or an error? The reason is that this sort of warnings are never desirable for end users and so should be considered programming errors.

In this case it seems OOB assignment should be an error? I think it's ok to have a lax `nth()` and strict `nth<-`. Or we could add an `oob = c("error", "ignore")` argument:

```r
nth(x, 2, oob = "ignore") <- 10
```

hadley commented on R/nth.R:
```diff
@@ -0,0 +1,108 @@
+  value <- vec_cast(value, x, x_arg = "value", to_arg = "x")
+
+  if (n == 0 || n > size || n < -size) {
+    warn("Unchanged `x` as `n` is out of bounds")
```
Yeah, I'm with @lionel- this should either be an error or we should silently expand the vector. I think erroring is fine here.

romainfrancois committed:
nth<-() stricter about bounds

romainfrancois commented:
Should both `n` tests be grouped together and maybe hint more, something like: 

```r
  if (!is_integerish(n, n = 1L, finite = TRUE)) {
    abort("`n` must be a single integer.")
  }
  if (n == 0 || n > size || n < -size) {
    abort(c(
      "`n` is out of bounds.",
      x = glue::glue("`n = {n}` cannot be used to modify a vector of size {size}."),
      i = glue::glue("Use a positive integer in [1,{size}] to select from the left."),
      i = glue::glue("Use a negative integer in [-{size}, -1] to select from the right.")
    ))
  }
```

This should maybe also abort if size is `0`

romainfrancois committed:
more detailed message
